### ISSUE TITLE
Screen Does Not Lock on MacOS

### ISSUE THREAD

**Issue by maltfield** [BusKill/buskill-app#35]
**Describe the bug**

A user has reported that their machine doesn't actually lock after arm & disconnect of the BusKill cable. It *does* make the screen go black, but it does *not* lock the machine (wiggling the mouse just makes the screen no-dark again without prompting for a password)

MacOS Monterey (12.6)

**To Reproduce**
1. Open the BusKill App
2. Connect the BusKill cable assembly to the computer
3. Arm the app
4. Disconnect the BusKill cable from the computer
5. Experience issue

**Expected behavior**

**Screenshots**

**Versions (please complete the following information):**
 - OS: MacOS running Monterey (12.6) (MacBook Pro)
 - BusKill: Reported for both v0.4.0 and v0.5.0

**Additional context**

**Comment by maltfield** [comment]
Just some quick googling around, and from what I'm reading, maybe the reason we couldn't reproduce this issue is that all of us paranoid BusKill hackers have set our machines to immediately lock the screen when the display goes off, but this user doesn't have their computer setup that way..

**Comment by maltfield** [comment]
I'm able to run the above builds, but I can't simulate a USB drive removal on my Mac Mini at MacStadium, so I've passed it to the user to see if they resolve the issue.

Also, if I could just execute the trigger function directly, this would help me to test at least the trigger functions themselves. So I created this feature request ticket:

 * https://github.com/BusKill/buskill-app/issues/38

**Comment by maltfield** [comment]
Unfortunately it appears that newer MacOS systems don't have `CGSession`, so the only way for BusKill to lock the screen is to put it to sleep.

And whether or not that means [a] a simple mouse-jiggle or [b] typing a password will unlock the system depends on your systems's settings

## Workaround

If you currently have this issue, please do the following:
1. Click the Apple logo in the top-left corner of your screen
2. Click "System Preferences"
3. Click "Security & Privacy"
4. Click the "General" Tab
5. Next to where it says "Require password <drop down menu> after sleep or screen saver begins", change the drop-down menu to `immediately`

**Comment by maltfield** [comment]
We just had another user report this, but I'm not yet sure if the above workaround solves the issue. If it does, then we should do the following to improve MacOS users UX:

1. Add a `Troubleshooting` section with an entry for `Screen does not lock on MacOS` to the documentation with easy-to-follow instructions (complete with screenshots) on how to do the above workaround steps on https://docs.buskill.in/
2. Add a section to the FAQ that links to [1] above
3. Update https://buskill.in/start with a link [1] above

**Comment by maltfield** [comment]
Here's a few more ways we can fallback and try try to lock the screen on MacOS, since CGSession availability appears to be spotty, and `pmset` depends on user's settings:
1. https://stackoverflow.com/questions/1976520/lock-screen-by-api-in-macos
2. https://github.com/nfirvine/osx-lock-on-remove-usb/blob/master/osx-lock-on-remove-usb.sh#L10
3. https://github.com/fluffypony/macos-yubikey-lock/blob/main/osx-lock-on-remove-usb.sh#L7

**Comment by maltfield** [comment]
I was successfully able to lock the screen by calling the (undocumented by Apple) `SACLockScreenImmediate()` function in the "login framework" library with the following python code:

```
import ctypes, ctypes.util

login = ctypes.CDLL( '/System/Library/PrivateFrameworks/login.framework/login' )
login.SACLockScreenImmediate();
```

**Comment by maltfield** [comment]
I created a new branch (`mac_lock_fix`) that adds two more fallback functions to attempt to lock the screen:

1. Executes `SACLockScreenImmediate()` as shown above
2. Opens the screensaver app (as fluffypony does above)

The latest build actually doesn't fall-back. It just tries to execute all 4 of them at once :sweat_smile: It can be found here:

 * https://github.com/BusKill/buskill-app/releases/tag/3292583182_mac

**Comment by maltfield** [comment]
I merged these two new lock attempts into the main `dev` branch, but with slightly different (less aggressive) logic:

1. First it tries to lock the screen with `CGSession`
2. If that fails, then it tries to lock the screen with `SACLockScreenImmediate()`
3. If that fails, then it executes both `ScreenSaverEngine.app` and `pmset displaysleepnow`

**Comment by maltfield** [comment]
I messaged both users who originally reported this bug to ask them to try the latest build, but neither responded :(

As far as I know, we've fixed this bug in v0.6.0

 * https://github.com/BusKill/buskill-app/releases/tag/v0.6.0
